packages/at_secondary_server/lib/src/verb/handler/pkam_verb_handler.dart

```diff
@@ -38,58 +39,116 @@ class PkamVerbHandler extends AbstractVerbHandler {
       publicKey = publicKeyData.data;
     }
 
-    // If there is no public key in the keystore then throw an exception
     if (publicKey == null || publicKey.isEmpty) {
-      response.data = 'failure';
-      response.isError = true;
-      response.errorMessage = 'pkam publickey not found';
       throw UnAuthenticatedException('pkam publickey not found');
     }
-    var isValidSignature = false;
+    var isValidSignature =
+        await _validateSignature(verbParams, sessionID, atSign, publicKey);
+    // authenticate if signature is valid
+    if (isValidSignature) {
+      atConnectionMetadata.isAuthenticated = true;
+      atConnectionMetadata.authType = pkamAuthType;
+      atConnectionMetadata.enrollmentId = enrollId;
+      response.data = 'success';
+    } else {
+      atConnectionMetadata.isAuthenticated = false;
+      logger.severe('pkam authentication failed');
+      throw UnAuthenticatedException('pkam authentication failed');
+    }
+  }
 
-    //retrieve stored secret using sessionid and atsign
-    var storedSecret = await keyStore.get('private:$sessionID$atSign');
-    storedSecret = storedSecret?.data;
+  @visibleForTesting
+  Future<ApkamVerificationResult> handleApkamVerification(
+      String enrollId, String atSign) async {
+    String enrollmentKey =
+        '$enrollId.$newEnrollmentKeyPattern.$enrollManageNamespace$atSign';
+    var enrollData = await keyStore.get(enrollmentKey);
+    final atData = enrollData.data;
+    final enrollDataStoreValue =
+        EnrollDataStoreValue.fromJson(jsonDecode(atData));
+    EnrollStatus enrollStatus =
+        EnrollStatus.values.byName(enrollDataStoreValue.approval!.state);
+
+    ApkamVerificationResult apkamResult = ApkamVerificationResult();
+    apkamResult.response = _getApprovalStatus(
+        enrollStatus, enrollId, enrollDataStoreValue.approval!.state);
+    if (apkamResult.response.isError) {
+      return apkamResult;
+    }
+    apkamResult.publicKey = enrollDataStoreValue.apkamPublicKey;
+    return apkamResult;
+  }
 
+  Response _getApprovalStatus(
+      EnrollStatus enrollStatus, enrollId, approvalState) {
+    Response response = Response();
+    switch (enrollStatus) {
+      case EnrollStatus.denied:
+        response.isError = true;
+        response.errorCode = 'AT0025';
+        response.errorMessage = 'enrollment_id: $enrollId is not approved |'
+            ' Status: $approvalState';
+        break;
+      case EnrollStatus.pending:
+        response.isError = true;
+        response.errorCode = 'AT0026';
+        response.errorMessage = 'enrollment_id: $enrollId is not approved |'
+            ' Status: $approvalState';
+        break;
+      case EnrollStatus.approved:
+        // do nothing when enrollment is approved
+        break;
+      case EnrollStatus.revoked:
+        response.isError = true;
+        response.errorCode = 'AT0027';
+        response.errorMessage = 'enrollment_id: $enrollId is not approved |'
```

Comment:
enrollment_id: $enrollId is revoked

Reply:
resolved as part of [3ade88d](https://github.com/atsign-foundation/at_server/pull/1520/commits/3ade88d17d2705baac492c17299f4aef2713acd7)